src/can.rs

```diff
@@ -0,0 +1,498 @@
+pub use embedded_hal_can::{self, Filter, Frame, Id, Receiver, Transmitter};
+
+use crate::gpio::gpioa;
+use crate::gpio::AF9;
+use crate::rcc::APB1;
+use crate::stm32;
+use arrayvec::ArrayVec;
+use nb;
+use nb::Error;
+
+use core::sync::atomic::{AtomicU8, Ordering};
+
+const EXID_MASK: u32 = 0b11111111111000000000000000000;
+const MAX_EXTENDED_ID: u32 = 0x1FFFFFFF;
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum CanId {
+    BaseId(u16),
+    ExtendedId(u32),
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct CanFrame {
+    pub id: CanId,
+    pub data: ArrayVec<[u8; 8]>,
+}
+
+pub enum FilterMode {
+    Mask,
+    List,
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum CanFilterData {
+    IdFilter(CanId),
+    MaskFilter(u16, u16),
+    ExtendedMaskFilter(u32, u32),
+    AcceptAll,
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct CanFilter {
+    data: CanFilterData,
+    index: Option<u8>,
+}
+
+static FILTER_INDEX: AtomicU8 = AtomicU8::new(0);
+
+pub struct Can {
```

Comment:
I would like to see more comments / documentation, especially on the public interfaces, like this struct :)